drivers/sensor/tsl2520/tsl2520.c

```diff
@@ -0,0 +1,198 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023 Carl Zeiss Meditec AG
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT ams_tsl2520
+
+#include "tsl2520.h"
+
+#include <stdlib.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/sys/util.h>
+
+#define TSL2520_CHIP_ID 0x5C
+
+LOG_MODULE_REGISTER(tsl2520, CONFIG_SENSOR_LOG_LEVEL);
+
+static int tsl2520_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	const struct tsl2520_config *cfg = dev->config;
+	struct tsl2520_data *data = dev->data;
+	int ret = 0;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL || chan == SENSOR_CHAN_LIGHT ||
+			chan == SENSOR_CHAN_IR);
+
+	if (chan == SENSOR_CHAN_ALL || chan == SENSOR_CHAN_LIGHT || chan == SENSOR_CHAN_IR) {
+		uint8_t reg_status;
+		uint16_t le16_buffer;
+
+		/* In order to update ALS Data Registers ALS_STATUS must be read first */
+		(void)i2c_reg_read_byte_dt(&cfg->i2c_spec, TSL2520_REG_STATUS, &reg_status);
+
+		ret = i2c_reg_read_byte_dt(&cfg->i2c_spec, TSL2520_REG_STATUS, &reg_status);
+		if (ret < 0) {
+			LOG_ERR("Failed reading chip status");
+			return -EIO;
+		}
+		ret = i2c_burst_read_dt(&cfg->i2c_spec, TSL2520_REG_ALS_DATA0_LOW,
+					(uint8_t *)&le16_buffer, sizeof(le16_buffer));
+		if (ret < 0) {
+			LOG_ERR("Could not fetch als sensor value");
+			return -EIO;
+		}
+		data->als_data0 = sys_le16_to_cpu(le16_buffer);
+		ret = i2c_burst_read_dt(&cfg->i2c_spec, TSL2520_REG_ALS_DATA1_LOW,
+					(uint8_t *)&le16_buffer, sizeof(le16_buffer));
+		if (ret < 0) {
+			LOG_ERR("Could not fetch als sensor value");
+			return -EIO;
+		}
+		data->als_data1 = sys_le16_to_cpu(le16_buffer);
+	}
+
+	return ret;
+}
+
+static int tsl2520_channel_get(const struct device *dev, enum sensor_channel chan,
+			       struct sensor_value *val)
+{
+	struct tsl2520_data *data = dev->data;
+
+	if (chan == SENSOR_CHAN_ALL || chan == SENSOR_CHAN_LIGHT || chan == SENSOR_CHAN_IR) {
+		val->val1 = data->als_data0;
+		val->val2 = data->als_data1;
+		return 0;
+	}
+
+	return -ENOTSUP;
+}
+
+static int tsl2520_setup(const struct device *dev)
+{
+	const struct tsl2520_config *cfg = dev->config;
+	int ret;
+	uint8_t tmp;
+	uint16_t temp;
+
+	ret = i2c_reg_write_byte_dt(&cfg->i2c_spec, TSL2520_ENABLE_ADDR, TSL2520_ENABLE_AEN_PON);
+	if (ret < 0) {
+		LOG_ERR("Failed to enable register");
+		return ret;
+	}
+
+	tmp = cfg->sample_time;
+	ret = i2c_reg_write_byte_dt(&cfg->i2c_spec, TSL2520_SAMPLE_TIME0, tmp);
+	if (ret < 0) {
+		LOG_ERR("Failed setting sample time");
+		return ret;
+	}
+
+	temp = cfg->als_nr_samples;
+	ret = i2c_burst_write_dt(&cfg->i2c_spec, TSL2520_ALS_NR_SAMPLES0, (uint8_t *)&temp,
+				 sizeof(temp));
```

Comment:
i2c_burst_write does not work correctly, so should be avoided (https://github.com/zephyrproject-rtos/zephyr/issues/20154).
Also you need to explicitly set the endianess of the data in the message, e.g:

```suggestion
	uint8_t buf[3] = {TSL2520_ALS_NR_SAMPLES0};
	sys_put_le16(cfg->als_nr_samples, &buf[1]);
	int ret = i2c_write_dt(&cfg->i2c_spec, buf, sizeof(buf));
```